worlds/satisfactory/StateLogic.py

```diff
@@ -0,0 +1,101 @@
+from typing import Optional
+from collections.abc import Iterable
+from BaseClasses import CollectionState
+from .GameLogic import GameLogic, Recipe, PowerInfrastructureLevel
+from .Options import SatisfactoryOptions
+from .CriticalPathCalculator import CriticalPathCalculator
+
+EventId: Optional[int] = None
+
+part_event_prefix = "Can Produce: "
+building_event_prefix = "Can Build: "
+
+
+class StateLogic:
+    player: int
+    options: SatisfactoryOptions
+    critical_path: CriticalPathCalculator
+    initial_unlocked_items: set[str]
+
+    def __init__(self, player: int, options: SatisfactoryOptions, critical_path: CriticalPathCalculator):
+        self.player = player
+        self.options = options
+        self.critical_path = critical_path
+
+    def has_recipe(self, state: CollectionState, recipe: Recipe) -> bool:
+        return state.has(recipe.name, self.player) or recipe.name in self.critical_path.implicitly_unlocked 
+    
+    def can_build(self, state: CollectionState, building_name: Optional[str]) -> bool:
+        return building_name is None or state.has(building_event_prefix + building_name, self.player)
+    
+    def can_build_any(self, state: CollectionState, building_names: Optional[Iterable[str]]) -> bool:
+        return building_names is None or \
+            state.has_any(map(self.to_building_event, building_names), self.player)
```

Comment:
I'm gonna use this as another example.

It'd be so much better to have a function that creates this rule efficiently:

```py
    def create_can_build_any_rule(self, building_names: Optional[Iterable[str]]) -> bool:
        if building_names is None:
            return lambda _: True
        
        building_event_names = map(self.to_building_event, building_names)

        return lambda state: state.has_any(building_event_names)
```

And then, whereever you are setting this rule, do:

```py
    can_build_any = state_logic.create_can_build_any_rule(building_names)
    set_rule(location, lambda: ... and can_build_any(state))
```

Reply:
Side note: If you do this, it would probably be best to save `lambda _: True` as a constant somewhere to save on memory